qadence/analog/addressing.py

```diff
@@ -0,0 +1,138 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from warnings import warn
+
+import sympy
+import torch
+from numpy import pi
+
+from qadence.parameters import Parameter, evaluate
+from qadence.types import StrEnum
+
+DEFAULT_MAX_AMPLITUDE = 2 * pi * 3
+DEFAULT_MAX_DETUNING = 2 * pi * 20
+
+
+class WeightConstraint(StrEnum):
+    """Supported types of constraints for addressing weights."""
+
+    NORMALIZE = "normalize"
+    """Normalize weights so that they sum up to 1."""
+
+    RESTRICT = "restrict"
+    """Restrict weight values to interval [0, 1]."""
+
+
+def sigmoid(x: torch.Tensor, a: float, b: float) -> sympy.Expr:
+    return 1.0 / (1.0 + sympy.exp(-a * (x + b)))
+
+
+@dataclass
+class AddressingPattern:
+    # number of qubits
+    n_qubits: int
+
+    # list of weights for fixed amplitude pattern that cannot be changed during the execution
+    weights_amp: dict[int, float | torch.Tensor | Parameter]
+
+    # list of weights for fixed detuning pattern that cannot be changed during the execution
+    weights_det: dict[int, float | torch.Tensor | Parameter]
+
+    # maximum amplitude can also be chosen as a variational parameter if needed
+    max_amp: float | torch.Tensor | Parameter = DEFAULT_MAX_AMPLITUDE
+
+    # maximum detuning can also be chosen as a variational parameter if needed
+    max_det: float | torch.Tensor | Parameter = DEFAULT_MAX_DETUNING
+
+    # weight constraint
+    weight_constraint: WeightConstraint = WeightConstraint.NORMALIZE
+
+    def _normalize_weights(self) -> None:
+        self.weights_amp = {
+            k: Parameter(v) if not isinstance(v, Parameter) else abs(v)
+            for k, v in self.weights_amp.items()
+        }
+        sum_weights_amp = sum(list(self.weights_amp.values()))
+        self.weights_amp = {k: v / sum_weights_amp for k, v in self.weights_amp.items()}
+
+        self.weights_det = {
+            k: Parameter(v) if not isinstance(v, Parameter) else abs(v)
+            for k, v in self.weights_det.items()
+        }
+        sum_weights_det = sum(list(self.weights_det.values()))
+        self.weights_det = {k: v / sum_weights_det for k, v in self.weights_det.items()}
```

Comment:
I would delete this function and add a condition in _restrict_weights

Reply:
Deleted this function.